orca-sql/src/main/kotlin/com/netflix/spinnaker/orca/sql/pipeline/persistence/SqlExecutionRepository.kt

```diff
@@ -267,30 +269,59 @@ class SqlExecutionRepository(
 
   override fun delete(type: ExecutionType, id: String) {
     doForeignAware(DeleteInterlinkEvent(type, id)) {
-      execution, dslContext ->
-      val correlationField = if (type == PIPELINE) "pipeline_id" else "orchestration_id"
-      val (ulid, _) = mapLegacyId(jooq, type.tableName, id)
+      _, dslContext ->
+      val (ulid, _) = mapLegacyId(dslContext, type.tableName, id)
 
-      val correlationId = jooq.select(field("id")).from("correlation_ids")
-        .where(field(correlationField).eq(ulid))
-        .limit(1)
-        .fetchOne()
-        ?.into(String::class.java)
-      val stageIds = jooq.select(field("id")).from(type.stagesTableName)
-        .where(field("execution_id").eq(ulid))
-        .fetch()
-        ?.into(String::class.java)?.toTypedArray()
-
-      if (correlationId != null) {
-        dslContext.delete(table("correlation_ids")).where(field("id").eq(correlationId)).execute()
-      }
-      if (stageIds != null) {
-        dslContext.delete(type.stagesTableName).where(field("id").`in`(*stageIds)).execute()
-      }
-      dslContext.delete(type.tableName).where(field("id").eq(ulid)).execute()
+      deleteInternal(dslContext, type, listOf(ulid))
+    }
+  }
+
+  /**
+   * Deletes given executions
+   * NOTE: this method explicitly does not check if the execution is foreign or not.
+   */
+  override fun delete(type: ExecutionType, idsToDelete: List<String>) {
+    jooq.transactional { tx ->
+      deleteInternal(tx, type, idsToDelete)
     }
   }
 
+  private fun deleteInternal(dslContext: DSLContext, type: ExecutionType, idsToDelete: List<String>) {
+    val correlationField = if (type == PIPELINE) "pipeline_id" else "orchestration_id"
+
+    dslContext
+      .delete(table("correlation_ids"))
+      .where(field(correlationField).`in`(*idsToDelete.toTypedArray()))
+      .execute()
+
+    dslContext
+      .delete(type.stagesTableName)
+      .where(field("execution_id").`in`(*idsToDelete.toTypedArray()))
+      .execute()
+
+    dslContext
+      .delete(type.tableName)
+      .where(field("id").`in`(*idsToDelete.toTypedArray()))
+      .execute()
+
+    var insertQueryBuilder = dslContext
+      .insertInto(table("deleted_executions"))
+      .columns(field("execution_id"), field("execution_type"), field("deleted_at"))
+
+    idsToDelete.forEach { id ->
+      insertQueryBuilder = insertQueryBuilder
+        .values(id, type.toString(), now())
+    }
+
+    insertQueryBuilder
+      .execute()
+
+    dslContext
+      .deleteFrom(table("deleted_executions"))
+      .where(field("deleted_at").lt(timestampSub(now(), 2, DatePart.DAY)))
```

Comment:
oh we only keep things deleted in the last 2 days in `deleted_executions`, am I reading this right? Sounds good but maybe we want to make that duration configurable

Reply:
will do

Reply:
actually... it adds a bunch of extra complexity that I am not sure it's worth it. I really can't imagine a situation where this would matter. If we delete too quickly (e.g. before the peer was able to peer them)... oh, well, when the peer is red/blacked it will do a full diff anyway and will pick up deletes. if the concern is that the table will fill up... i don't think that's a problem. even if there is an execution that is deleted every second that's 172k rows which is of course a bit crazy, but not really a big deal (since the peer will only read since its known cursor). I will reduce the retentio to 1 day